File: src/main/java/com/ada/conversation/context/CompactedPromptContextItem.java
package com.ada.conversation.context;

import com.ada.conversation.application.dto.*;
import com.ada.shared.observability.*;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

@Component
@Order(60)
@MeasuredContextItem("compacted_prompt")
@RequiredArgsConstructor
public class CompactedPromptContextItem implements ContextItem {
  private final TokenUsageEstimator estimator;

  @Value("${ada.context.max-tokens:12000}")
  private long max;

  public LlmContentComponent component() {
    return LlmContentComponent.COMPACTED_PROMPT;
  }

  public ContextState apply(ChatRequest r, ContextState c) {
    if (estimator.estimate(c) <= max) return c;
    var keep =
        c.messages().stream().filter(m -> m.component() == LlmContentComponent.SYSTEM).toList();
    var summary =
        c.messages().stream()
            .filter(m -> m.component() != LlmContentComponent.SYSTEM)
            .map(m -> m.role().wireName() + ": " + m.content())
            .collect(java.util.stream.Collectors.joining("\n"));
    var all = new java.util.ArrayList<>(keep);
    all.add(
        new LlmMessage(
            LlmMessageRole.SYSTEM, "Previous context summary:\n" + summary, component()));
    return new ContextState(all, c.tools());
  }
}

File: src/main/kotlin/com/ada/conversation/context/CompactedPromptContextItem.kt
package com.ada.conversation.context

import com.ada.conversation.application.dto.ChatRequest
import com.ada.conversation.application.dto.ContextState
import com.ada.conversation.application.dto.LlmMessage
import com.ada.conversation.application.dto.LlmMessageRole
import com.ada.shared.observability.TokenUsageEstimator
import org.springframework.beans.factory.annotation.Value
import com.ada.conversation.application.dto.LlmContentComponent
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import com.ada.shared.observability.MeasuredContextItem

@Component
@Order(60)
@MeasuredContextItem("compacted_prompt")
class CompactedPromptContextItem(
    private val tokenUsageEstimator: TokenUsageEstimator,
    @Value("\${ada.context.max-tokens:12000}") private val maxTokens: Long,
) : ContextItem {
    override val component = LlmContentComponent.COMPACTED_PROMPT

    override fun apply(request: ChatRequest, current: ContextState): ContextState {
        if (tokenUsageEstimator.estimate(current) <= maxTokens) return current

        val systemMessages = current.messages.filter { it.component == LlmContentComponent.SYSTEM }
        val messagesToCompact = current.messages.filterNot { it.component == LlmContentComponent.SYSTEM }
        val summary = messagesToCompact.joinToString("\n") { "${it.role.wireName()}: ${it.content}" }
        val compactedMessage = LlmMessage(
            LlmMessageRole.SYSTEM,
            "Previous context summary:\n$summary",
            component,
        )
        return current.copy(messages = systemMessages + compactedMessage)
    }
}
